domain/stub/service.go

```diff
@@ -110,3 +126,45 @@ WHERE name IN ($units[:])
 	}
 	return err
 }
+
+// CloudSpec returns the cloud spec for the model.
+func (s *StubService) CloudSpec(ctx context.Context) (cloudspec.CloudSpec, error) {
+	modelSt := modelstate.State{StateBase: s.controllerState}
```

Comment:
It would be better if we used the model database for the model details, but I'm not going to stop this from landing based on this.

For people following along, this is on the stub service, because the information to get a cloudspec is found deep down in the provider tracker.

https://github.com/juju/juju/blob/3b5991bb77d2249bfcff6b5b23acbd2eec6ad356/internal/worker/providertracker/trackerworker.go#L346-L358

We should lift this method out, so both the provider and the uniter, are using the same code. That's a future us problem.

Reply:
I am not sure we could use the model database to build a cloud spec on its own.

Reply:
Added a new method `GetModelCloudRegionAndCredential` to model state so that model db is used to get the model details.